cadence/contracts/bridge/FlowEVMBridge.cdc

```diff
@@ -0,0 +1,405 @@
+        self.deployLockerContract(forType: type)
+    }
+
+    /// Onboards a given ERC721 to the bridge. Since we're onboarding by EVM Address, the asset must be defined in a
+    /// third-party EVM contract. Attempting to onboard a bridge-defined asset will result in an error as onboarding is
+    /// not required
+    ///
+    /// @param address: The EVMAddress of the ERC721 or ERC20 to be onboarded
+    /// @param tollFee: Fee paid for onboarding
+    ///
+    access(all) fun onboardByEVMAddress(_ address: EVM.EVMAddress, tollFee: @FlowToken.Vault) {
+        pre {
+            tollFee.balance >= FlowEVMBridgeConfig.fee: "Insufficient fee paid"
+        }
+        FlowEVMBridgeUtils.depositTollFee(<-tollFee)
+        assert(
+            FlowEVMBridgeUtils.isValidEVMAsset(evmContractAddress: address),
+            message: "Target contract address is not a valid asset type"
+        )
+        assert(
+            self.evmAddressRequiresOnboarding(address: address) == true,
+            message: "Onboarding is not needed for this contract"
+        )
+        self.deployDefiningContract(evmContractAddress: address)
+    }
+
+    /// Public entrypoint to bridge NFTs from Flow to EVM - cross-account bridging supported (e.g. straight to EOA)
+    ///
+    /// @param token: The NFT to be bridged
+    /// @param to: The NFT recipient in FlowEVM
+    /// @param tollFee: The fee paid for bridging
+    ///
+    access(all) fun bridgeNFTToEVM(token: @{NonFungibleToken.NFT}, to: EVM.EVMAddress, tollFee: @FlowToken.Vault) {
+        pre {
+            tollFee.balance >= FlowEVMBridgeConfig.fee: "Insufficient fee paid"
+            token.isInstance(Type<@{FungibleToken.Vault}>()) == false: "Mixed asset types are not yet supported"
+            self.typeRequiresOnboarding(token.getType()) == false: "NFT must first be onboarded"
+        }
+        if FlowEVMBridgeUtils.isFlowNative(type: token.getType()) {
+            self.bridgeFlowNativeNFTToEVM(token: <-token, to: to, tollFee: <-tollFee)
+        } else {
+            // TODO: EVM-native NFT path
+            // self.bridgeEVMNativeNFTToEVM(token: <-token, to: to, tollFee: <-tollFee)
+            destroy <- token
+            destroy <- tollFee
+        }
+    }
+
+    /// Public entrypoint to bridge NFTs from EVM to Flow
+    ///
+    /// @param caller: The caller executing the bridge - must be passed to check EVM state pre- & post-call in scope
+    /// @param calldata: Caller-provided approve() call, enabling contract COA to operate on NFT in EVM contract
+    /// @param id: The NFT ID to bridged
+    /// @param evmContractAddress: Address of the EVM address defining the NFT being bridged - also call target
+    /// @param tollFee: The fee paid for bridging
+    ///
+    /// @returns The bridged NFT
+    ///
+    access(all) fun bridgeNFTFromEVM(
+        caller: &EVM.BridgedAccount,
```

Comment:
`BridgedAccount` is the resource (recently renamed to `CadenceOwnedAccount`) and `EVMAddress` is a struct containing the 20 bytes of the underlying address.

We need the caller's `BridgedAccount` here because the NFT being bridged is owned by them and the `approve` call must come from them. In an ideal world, we wouldn't be passing a reference to the fully-entitled resource, but for a) I'm waiting on [Cadence 1.0 updates to the EVM contract](https://github.com/onflow/flow-go/pull/5390) to restrict this a bit and b) in future iterations we could think about wrapping the `BridgedAccount` in a restricted resource that would limit the calls that can be made on the underlying `BridgedAccount` and passing a reference/Capability to that.

Especialy in a world with self-rolled bridges, the wrapper sounds like a much safer best practice as passing around a `BridgedAccount` is akin to passing around an `AuthAccount`

Reply:
Makes sense